Adding progress bar support for reporter backends (#14083)
conda/conda: docs/source/dev-guide/plugins/reporter_backends.rst

Reporter Backends
=================

A reporter backend is a plugin hook that allows you to change the output displayed
for existing conda commands. Internally, conda uses this to differentiate between
displaying normal, "console" output or displaying "json" output. An example of each
of these reporter backend types is shown below:

.. code-block:: shell

   # "console" output
   conda info --envs

   # conda environments:
   #
   base                     * /home/user/conda/
   env-1                      /home/user/.conda/envs/env-1

   # "json" output (output has been truncated)
   conda info --envs --json
   {
   ...
     "envs": [
       "/home/user/conda/",
       "/home/user/.conda/envs/env-1"
     ]
   ...
   }

Via its plugins system, conda allows you to define entirely new reporter backends. This gives

> you an extreme amount of flexibility when it comes to modifying how conda looks

Reviewer: ```suggestion
you a substantial amount of flexibility when it comes to modifying how conda looks
```

        """

        def detail_view(self, data):
            return pformat(data)

        def envs_list(self, data):
            formatted_data = pformat(data)
            return f"Environments: {formatted_data}"

        def progress_bar(self, description, io_context_manager) -> ProgressBarBase:
            """Returns our custom progress bar implementation"""
            return PprintProgressBar(description, io_context_manager)

    class PprintProgressBar(ProgressBarBase):
        """
        Blank implementation of ProgressBarBase which does nothing.
        """

        def update_to(self, fraction) -> None:
            pass

        def refresh(self) -> None:
            pass

        def close(self) -> None:
            pass

    @plugins.hookimpl
    def conda_reporter_backends():
        yield CondaReporterBackend(
            name="pprint",
            description="Reporter backend based on the pprint module",
            renderer=PprintReporterRenderer,
        )

Below is a summary of everything we've defined:

Defining ``ReporterRendererBase``
---------------------------------

The first class we define, ``PprintReporterRenderer``, is a subclass of

> :class:`~conda.plugins.types.ReporterRendererBase`. The base class is an abstract base class which requires us

Reviewer: ```suggestion
:class:`~conda.plugins.types.ReporterRendererBase`. The base class is an abstract base class which requires us to
```
